freeipa/src/main/java/com/sequenceiq/freeipa/service/recipe/FreeIpaRecipeService.java

```diff
@@ -105,6 +107,21 @@ public void deleteRecipes(Long stackId) {
         freeIpaStackRecipeRepository.deleteFreeIpaStackRecipesByStackId(stackId);
     }
 
+    public void attachRecipes(Long stackId, Collection<String> newRecipes) {
+        Set<String> existingRecipes = getRecipeNamesForStack(stackId);
+        recipeCrnListProviderService.validateRequestedRecipesExistsByName(newRecipes);
+        List<String> recipesToSave = newRecipes.stream().filter(newRecipe -> !existingRecipes.contains(newRecipe)).collect(Collectors.toList());
+        saveRecipes(recipesToSave, stackId);
+    }
+
+    public void detachRecipes(Long stackId, Collection<String> recipesToRemove) {
+        Collection<String> existingRecipes = getRecipeNamesForStack(stackId);
+        recipesToRemove.stream().filter(recipe -> !existingRecipes.contains(recipe)).findFirst().ifPresent(recipe -> {
```

Comment:
It would be good to collect all non-existing recipes and throw exception which contains all of them and not just the first one, because it can be misleading.
And please add logging.